### ISSUE TITLE
1.0.0-RC2: using arg as argument name creates IR error

### ISSUE THREAD

**Issue by aappddeevv** [scala-js/scala-js#3933]
I cannot pinpoint the issue here but this code:

```scala
  val sfc = SFC1[Props] { arg =>
    val cond = arg.summary.facets.headOption.map(f => convertCondition(f.label)).orUndefined
    val props = new ReactWeatherDisplayProps {
      temperature = arg.summary.temp
      currentTemperature = arg.summary.temp
      width = 150
      height = 150
      condition = cond
      currentCondition = cond
    }
    div(new DivProps {
      className = "daily"
      style = layout
    })(
      Label(labelProps)(arg.summary.dateStr),
      ReactWeatherDisplay(props)
    )
  }
```

gives an IR error

```sh
[info] Fast optimizing /home/me/proj/workspace/scalajs-reaction/examples/target/scala-2.13/Scala.js
[error] file:/home/me/proj/workspace/scalajs-reaction/examples/src/main/scala/atmoache/atmoache.scala(134:21:VarDef): Duplicate local variable name arg$1.
[error] There were 1 IR checking errors.
[error] (examples / Compile / fastOptJS) There were 1 IR checking errors.
```

If a change arg to argx it compiles fine.

**Comment by gzm0** [comment]
Could you post the output of `examples/scalajsp fully.qualified.class.Name` of the class containing this?

**Comment by aappddeevv** [comment]
I'm glad I learned about scalajsp today!

```javascript
class ttg.examples.atmoache.DailyWeatherSummary extends java.lang.Object {
  static def sfc;Lreact.SFC1(): react.SFC1 = {
    mod:ttg.examples.atmoache.DailyWeatherSummary$.sfc;Lreact.SFC1()
  }
  static def apply;Lttg.examples.atmoache.package$Daily;Ljava.lang.String;Lreact.package$ReactElement(s: ttg.examples.atmoache.package$Daily, k: java.lang.String): any = {tions 0s
    mod:ttg.examples.atmoache.DailyWeatherSummary$.apply;Lttg.examples.atmoache.package$Daily;Ljava.lang.String;Lreact.package$ReactElement(s, k)
  } => jss / scalacOptions 0s
  static def apply;Lttg.examples.atmoache.DailyWeatherSummary$Props;Lreact.package$ReactElement(props: any): any = {
    mod:ttg.examples.atmoache.DailyWeatherSummary$.apply;Lttg.examples.atmoache.DailyWeatherSummary$Props;Lreact.package$ReactElement(props)
  }
  static def labelProps;Lfabric.components.Label$Props(): any = {
    mod:ttg.examples.atmoache.DailyWeatherSummary$.labelProps;Lfabric.components.Label$Props()
  }
  static def layout;Lreact.vdom.StyleAttr(): any = {
    mod:ttg.examples.atmoache.DailyWeatherSummary$.layout;Lreact.vdom.StyleAttr()
  }
  static def convertCondition;Ljava.lang.String;Ljava.lang.String(apiCondition: java.lang.String): java.lang.String = {
    mod:ttg.examples.atmoache.DailyWeatherSummary$.convertCondition;Ljava.lang.String;Ljava.lang.String(apiCondition)
  }
  static def Name;Ljava.lang.String(): java.lang.String = {
    mod:ttg.examples.atmoache.DailyWeatherSummary$.Name;Ljava.lang.String()
  }
[success] Total time: 17 s, completed Jan 19, 2020, 3:24:12 PM
sbt:root> 
```

**Comment by aappddeevv** [comment]
Come to think of it, I'm glad I've had such few problems with scala.js before.

**Comment by gzm0** [comment]
Minimum reproducible example:

```scala
object HelloWorld {
  def main(args: Array[String]): Unit = {
    foo { arg =>
      new js.Object {
        println(arg)
      }
    }
  }

  def foo(f: Int => Unit): Unit = ()
}
```

Scalajsp

```
module class helloworld.HelloWorld$ extends java.lang.Object {
  def main;[Ljava.lang.String;V(args: java.lang.String[]) {
    this.foo;Lscala.Function1;V(new scala.scalajs.runtime.AnonFunction1().<init>;Lscala.scalajs.js.Function1;V((arrow-lambda<this{this}: helloworld.HelloWorld$ = this>(arg$2{arg}: any) = {
      val arg: int = arg$2.asInstanceOf[int];
      this.helloworld.HelloWorld$::private::$anonfun$main$1;I;V(arg);
      (void 0)
    })))
  }
  def foo;Lscala.Function1;V(f: scala.Function1) {
    /*<skip>*/
  }
  @hints(1) private def $anonfun$main$1;I;V(arg: int) {
    (arrow-lambda<>(super: any, arg$1: any) = {
      val prep0: int = arg$1.asInstanceOf[int];
      val arg$1: int = prep0;
      val this{this}: any = {};
      mod:scala.Predef$.println;Ljava.lang.Object;V(arg$1);
      (void 0);
      this
    })(constructorOf[scala.scalajs.js.Object], arg)
  }
  constructor def <init>;V() {
    this.java.lang.Object::<init>;V();
    mod:helloworld.HelloWorld$<-this
  }
}
```

**Comment by gzm0** [comment]
The problem seems to be that the code generating anonymous JS classes does not handle names creation properly: `arg$1` is both an argument to the lambda and a local variable.

**Comment by gzm0** [comment]
The problem seems to be that both the class capture of `arg` and the exported constructor claim the name `arg$1`. They do so for different reasons: The class capture appends `$1` to the captured name (where `1` is probably the argument index). The exported constructor simply enumerates its arguments.

**Comment by sjrd** [comment]
The same minimization is also problematic in Scala.js 0.6.x. It just wasn't discovered because 0.6.x does not enable the IR checker by default, and the optimizer "fixes up" duplicated var names like that.

The 0.6.x IR is:
```scala
module class Lhelloworld_HelloWorld$ extends O {
  def main__AT__V(args: T[]) {
    this.foo__F1__V(new sjsr_AnonFunction1().init___sjs_js_Function1((lambda<$this: Lhelloworld_HelloWorld$ = this>(arg$2: any) = {
      val arg: int = arg$2.asInstanceOf[I];
      $this.$$anonfun$main$1__p1__I__V(arg);
      (void 0)
    })))
  }
  def foo__F1__V(f: F1) {
    /*<skip>*/
  }
  @hints(1) def $$anonfun$main$1__p1__I__V(arg: int) {
    (lambda<>(arg$1: any) = {
      val prep0: int = arg$1.asInstanceOf[I];
      val arg$1: int = prep0;
      val $this: any = {};
      mod:s_Predef$.println__O__V(arg$1);
      $this
    })(arg)
  }
  def init___() {
    this.O::init___();
    mod:Lhelloworld_HelloWorld$<-this
  }
}
```

**Comment by sjrd** [comment]
Minimizing further ... it can be reproduced without any lambda. It is enough to have a parameter of a `def` named `arg`:
```scala
object HelloWorld {
  def main(args: Array[String]): Unit = {
    bar(5)
  }

  def bar(arg: Int): Any = {
    new js.Object {
      println(arg)
    }
  }
}
```
IR in 0.6.x:
```scala
module class Lhelloworld_HelloWorld$ extends O {
  def main__AT__V(args: T[]) {
    this.bar__I__O(5)
  }
  def bar__I__O(arg: int): any = {
    (lambda<>(arg$1: any) = {
      val prep0: int = arg$1.asInstanceOf[I];
      val arg$1: int = prep0;
      val $this: any = {};
      mod:s_Predef$.println__O__V(arg$1);
      $this
    })(arg)
  }
  def init___() {
    this.O::init___();
    mod:Lhelloworld_HelloWorld$<-this
  }
}
```
Since it is reproducible in 0.6.x, it also has nothing to do with class captures (a concept that was introduced in 1.x). It is only a anonymous JS object's constructor's parameter that clashes with an enclosing local var name `arg`.

**Comment by sjrd** [comment]
An even simpler minimization:
```scala
import scala.scalajs.js

object HelloWorld {
  def main(args: Array[String]): Unit = {
    new Bar(5)
  }
}

class Bar(arg$1: Int) extends js.Object {
  println(arg$1)
}
```
With a named, top-level `Bar`. The IR of `Bar` is
```scala
js class Lhelloworld_Bar extends sjs_js_Object {
  def "constructor"(arg$1: any) {
    val prep0: int = arg$1.asInstanceOf[I];
    val arg$1: int = prep0;
    super();
    mod:s_Predef$.println__O__V(arg$1)
  }
}
```

Granted, no one would actually write their argument `arg$1` by hand. But this highlights that the issue is already happening when we build the constructor of a JS class. It is not related to anonymous JS classes at all.

**Comment by sjrd** [comment]
Fixed in 787033acdc8965c6f1f94f803cd5aaf9aa771ff5.
